### DISCUSSION TITLE
Cross file changes

### DISCUSSION THREAD

**Discussion by sakura-sukii** [GumTreeDiff/gumtree#370] [Q&A]
How can I use Gumtreediff to find cross file changes?
There may have some code be deleted in src.py and added in dst.py. Something like a 'move', can Gumtree deel with these conditions?

**Comment by jrfaller** [comment]
This is not something that is implemented, unfortunately...

**Reply by sakura-sukii** [reply]
I noticed that 'cross file change' is your ongoing work. Is there something difficult to implement, or just need time? What is the crucial problem

**Comment by jrfaller** [comment]
Hi!

Unfortunately, this is not something that I am currently working on. I don't think that it's impossible to implement, although it will require careful thinking to avoid a combinatorial explosion. I think the key point is to perform matching first within-file and try cross-file matching in a second round. Maybe a new hyperparameter to not be too aggressive in the cross-file matching. Finally, maybe an index would help to speed-up the process.
Another thing to design is also a view for such a multifile diff that scales reasonably well.
Given all these challenges, I think coming up with a good working prototype would be a good research contribution.
I believe that @tsantalis has implemented a cross-file differ, it would be nice to look into what he did.

Cheers!

**Comment by tsantalis** [comment]
@sakura-sukii 
Our tool [RefactoringMiner](https://github.com/tsantalis/RefactoringMiner) can handle file renames/moves and also supports code moves between files. Unfortunately, it supports only Java for the moment.

However, in our experiments, we implemented the Staged-Tree-Matching algorithm by Fujimoto et al., which has a second stage trying to match the remaining unmatched AST nodes between different files.
https://dl.acm.org/doi/10.1145/3387904.3389289

The implementation can be found at 
https://github.com/pouryafard75/DiffBenchmark/blob/master/src/main/java/benchmark/generators/tools/runners/hacks/interfile/StagedTreeMatching.java

@pouryafard75 can give you more information

**Reply by Timmmm** [reply]
Phabricator/Phorge also supports this to some extent in its diff pages. Kind of blows my mind how little attention this problem has received.

**Reply by tsantalis** [reply]
@Timmmm 
BTW, RefactoringMiner now supports more programming languages, including Java, Python, Kotlin, TypeScript. For all these languages it can detect moves between files.

**Reply by Timmmm** [reply]
Unfortunately I'm using C++. :-/

**Reply by tsantalis** [reply]
C++ is in our plans. Stay tuned.

**Reply by tsantalis** [reply]
@Timmmm 
C++ is now supported by RefactoringMiner. It would be great to have your feedback.
